NetKVM/ProtocolService/ProtocolService.cpp

```diff
@@ -29,37 +29,728 @@
 
 #include "stdafx.h"
 
-class CProtocolServiceImplementation : public CServiceImplementation
+/*
+    The protocol service replaces and extends the notification object.
+    Unlike previous implementation when the VIOPROT was enabled only
+    on especially defined set of adapters, now the VIOPROT is installed and
+    enabled on all the ethernet adapters. The service is responsible for:
+    - disable all other protocols/filters only when needed (when the adapter
+      has the same MAC as virtio-net adapter)
+    - enable all other protocols when the protocol is uninstalled
+    - make standby netkvm adapter functional after several seconds if the VF
+      is not present and reconnect via VF when the VF comes
+    - enable all other protocols on VF when the netkvm is disabled
+*/
+
+class CNetCfg
 {
 public:
-    CProtocolServiceImplementation() : CServiceImplementation(_T("netkvmp")) {}
+    CNetCfg()
+    {
+        hr = CoCreateInstance(CLSID_CNetCfg, NULL, CLSCTX_SERVER, IID_INetCfg, (LPVOID*)&m_NetCfg);
+        if (!m_NetCfg) {
+            return;
+        }
+        hr = m_NetCfg->QueryInterface(IID_INetCfgLock, (LPVOID*)&m_NetCfgLock);
+        if (!m_NetCfgLock) {
+            return;
+        }
+        m_Usable = m_NetCfgLock->AcquireWriteLock(5000, L"netkvmp service", NULL) == S_OK;
+        if (m_Usable) {
+            hr = m_NetCfg->Initialize(NULL);
+            if (hr != S_OK) {
+                m_NetCfgLock->ReleaseWriteLock();
+                m_Usable = false;
+            }
+        }
+    }
+    ~CNetCfg()
+    {
+        if (!m_Usable) {
+            return;
+        }
+        if (m_Modified)
+        {
+            Log("Applying binding changes");
+            m_NetCfg->Apply();
+        }
+        m_NetCfgLock->ReleaseWriteLock();
+        m_NetCfg->Uninitialize();
+    }
+    bool Usable() const { return m_Usable; }
+    void EnableVioProtOnly(const CString& Name, bool Enable)
```

Comment:
Nice to have: break the function into smaller functions.

Reply:
Splitted